// @pnpm/default-reporter/lib/reporterForClient/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const operators_1 = require("rxjs/operators");
const reportBigTarballsProgress_1 = __importDefault(require("./reportBigTarballsProgress"));
const reportContext_1 = __importDefault(require("./reportContext"));
const reportDeprecations_1 = __importDefault(require("./reportDeprecations"));
const reportHooks_1 = __importDefault(require("./reportHooks"));
const reportInstallChecks_1 = __importDefault(require("./reportInstallChecks"));
const reportLifecycleScripts_1 = __importDefault(require("./reportLifecycleScripts"));
const reportMisc_1 = __importDefault(require("./reportMisc"));
const reportProgress_1 = __importDefault(require("./reportProgress"));
const reportRequestRetry_1 = __importDefault(require("./reportRequestRetry"));
const reportScope_1 = __importDefault(require("./reportScope"));
const reportSkippedOptionalDependencies_1 = __importDefault(require("./reportSkippedOptionalDependencies"));
const reportStats_1 = __importDefault(require("./reportStats"));
const reportSummary_1 = __importDefault(require("./reportSummary"));
function default_1(log$, opts) {
    var _a, _b, _c, _d;
    const width = (_b = (_a = opts.width) !== null && _a !== void 0 ? _a : process.stdout.columns) !== null && _b !== void 0 ? _b : 80;
    const cwd = (_d = (_c = opts.pnpmConfig) === null || _c === void 0 ? void 0 : _c.dir) !== null && _d !== void 0 ? _d : process.cwd();
    const throttle = typeof opts.throttleProgress === 'number' && opts.throttleProgress > 0
        ? operators_1.throttleTime(opts.throttleProgress, undefined, { leading: true, trailing: true })
        : undefined;
    const outputs = [
        reportProgress_1.default(log$, {
            cwd,
            throttle,
        }),
        reportLifecycleScripts_1.default(log$, {
            appendOnly: opts.appendOnly === true || opts.streamLifecycleOutput,
            cwd,
            width,
        }),
        reportDeprecations_1.default(log$.deprecation, { cwd, isRecursive: opts.isRecursive }),
        reportMisc_1.default(log$, {
            config: opts.config,
            cwd,
            logLevel: opts.logLevel,
            zoomOutCurrent: opts.isRecursive,
        }),
        ...reportStats_1.default(log$, {
            cmd: opts.cmd,
            cwd,
            isRecursive: opts.isRecursive,
            width,
        }),
        reportInstallChecks_1.default(log$.installCheck, { cwd }),
        reportRequestRetry_1.default(log$.requestRetry),
        reportScope_1.default(log$.scope, { isRecursive: opts.isRecursive, cmd: opts.cmd }),
        reportSkippedOptionalDependencies_1.default(log$.skippedOptionalDependency, { cwd }),
        reportHooks_1.default(log$.hook, { cwd, isRecursive: opts.isRecursive }),
        reportContext_1.default(log$, { cwd }),
    ];
    if (!opts.appendOnly) {
        outputs.push(reportBigTarballsProgress_1.default(log$));
    }
    if (!opts.isRecursive) {
        outputs.push(reportSummary_1.default(log$, {
            cwd,
            pnpmConfig: opts.pnpmConfig,
        }));
    }
    return outputs;
}
exports.default = default_1;
