// @pnpm/modules-cleaner/lib/removeDirectDependency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_loggers_1 = require("@pnpm/core-loggers");
const remove_bins_1 = require("@pnpm/remove-bins");
const path = require("path");
async function removeDirectDependency(dependency, opts) {
    const dependencyDir = path.join(opts.modulesDir, dependency.name);
    const results = await Promise.all([
        remove_bins_1.removeBinsOfDependency(dependencyDir, opts),
        !opts.dryRun && remove_bins_1.removeBin(dependencyDir),
    ]);
    const uninstalledPkg = results[0];
    if (!opts.muteLogs) {
        core_loggers_1.rootLogger.debug({
            prefix: opts.rootDir,
            removed: {
                dependencyType: dependency.dependenciesField === 'devDependencies' && 'dev' ||
                    dependency.dependenciesField === 'optionalDependencies' && 'optional' ||
                    dependency.dependenciesField === 'dependencies' && 'prod' ||
                    undefined,
                name: dependency.name,
                version: uninstalledPkg === null || uninstalledPkg === void 0 ? void 0 : uninstalledPkg.version,
            },
        });
    }
}
exports.default = removeDirectDependency;
