// @pnpm/plugin-commands-installation/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.update = exports.unlink = exports.remove = exports.prune = exports.link = exports.install = exports.add = void 0;
const add = __importStar(require("./add"));
exports.add = add;
const install = __importStar(require("./install"));
exports.install = install;
const link = __importStar(require("./link"));
exports.link = link;
const prune = __importStar(require("./prune"));
exports.prune = prune;
const remove = __importStar(require("./remove"));
exports.remove = remove;
const unlink = __importStar(require("./unlink"));
exports.unlink = unlink;
const update = __importStar(require("./update"));
exports.update = update;
