// @pnpm/local-resolver/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("@pnpm/error"));
const read_project_manifest_1 = require("@pnpm/read-project-manifest");
const parsePref_1 = __importDefault(require("./parsePref"));
const fs = require("graceful-fs");
const ssri = require("ssri");
/**
 * Resolves a package hosted on the local filesystem
 */
async function resolveLocal(wantedDependency, opts) {
    var _a;
    const spec = parsePref_1.default(wantedDependency.pref, opts.projectDir, (_a = opts.lockfileDir) !== null && _a !== void 0 ? _a : opts.projectDir);
    if (!spec)
        return null;
    if (spec.type === 'file') {
        return {
            id: spec.id,
            normalizedPref: spec.normalizedPref,
            resolution: {
                integrity: await getFileIntegrity(spec.fetchSpec),
                tarball: spec.id,
            },
            resolvedVia: 'local-filesystem',
        };
    }
    let localDependencyManifest;
    try {
        localDependencyManifest = await read_project_manifest_1.readProjectManifestOnly(spec.fetchSpec);
    }
    catch (internalErr) {
        switch (internalErr.code) {
            case 'ENOTDIR': {
                throw new error_1.default('NOT_PACKAGE_DIRECTORY', `Could not install from "${spec.fetchSpec}" as it is not a directory.`);
            }
            case 'ENOENT': {
                throw new error_1.default('DIRECTORY_HAS_NO_PACKAGE_JSON', `Could not install from "${spec.fetchSpec}" as it does not contain a package.json file.`);
            }
            default: {
                throw internalErr;
            }
        }
    }
    return {
        id: spec.id,
        manifest: localDependencyManifest,
        normalizedPref: spec.normalizedPref,
        resolution: {
            directory: spec.dependencyPath,
            type: 'directory',
        },
        resolvedVia: 'local-filesystem',
    };
}
exports.default = resolveLocal;
async function getFileIntegrity(filename) {
    return (await ssri.fromStream(fs.createReadStream(filename))).toString();
}
