// @pnpm/list/lib/getPkgInfo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const readPkg_1 = __importDefault(require("./readPkg"));
const path = require("path");
async function getPkgInfo(pkg) {
    var _a;
    let manifest;
    try {
        manifest = await readPkg_1.default(path.join(pkg.path, 'node_modules', pkg.name, 'package.json'));
    }
    catch (err) {
        // This will probably never happen
        manifest = {
            description: '[Could not find additional info about this dependency]',
        };
    }
    return {
        alias: pkg.alias,
        from: pkg.name,
        version: pkg.version,
        resolved: pkg.resolved,
        description: manifest.description,
        homepage: manifest.homepage,
        repository: (_a = (manifest.repository && (typeof manifest.repository === 'string' ? manifest.repository : manifest.repository.url))) !== null && _a !== void 0 ? _a : undefined,
    };
}
exports.default = getPkgInfo;
