// @pnpm/list/lib/renderJson.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toJsonResult = void 0;
const types_1 = require("@pnpm/types");
const getPkgInfo_1 = __importDefault(require("./getPkgInfo"));
const R = require("ramda");
const sortPackages = R.sortBy(R.path(['pkg', 'alias']));
async function default_1(pkgs, opts) {
    const jsonArr = await Promise.all(pkgs.map(async (pkg) => {
        var _a;
        const jsonObj = {
            name: pkg.name,
            version: pkg.version,
        };
        for (const dependenciesField of [...types_1.DEPENDENCIES_FIELDS.sort(), 'unsavedDependencies']) {
            if ((_a = pkg[dependenciesField]) === null || _a === void 0 ? void 0 : _a.length) {
                jsonObj[dependenciesField] = await toJsonResult(pkg[dependenciesField], { long: opts.long });
            }
        }
        return jsonObj;
    }));
    return JSON.stringify(jsonArr, null, 2);
}
exports.default = default_1;
async function toJsonResult(entryNodes, opts) {
    const dependencies = {};
    await Promise.all(sortPackages(entryNodes).map(async (node) => {
        var _a;
        const subDependencies = await toJsonResult((_a = node.dependencies) !== null && _a !== void 0 ? _a : [], opts);
        const dep = opts.long
            ? await getPkgInfo_1.default(node)
            : {
                alias: node.alias,
                from: node.name,
                version: node.version,
                resolved: node.resolved,
            };
        if (Object.keys(subDependencies).length) {
            dep['dependencies'] = subDependencies;
        }
        if (!dep.resolved) {
            delete dep.resolved;
        }
        delete dep.alias;
        dependencies[node.alias] = dep;
    }));
    return dependencies;
}
exports.toJsonResult = toJsonResult;
