// @pnpm/read-projects-context/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lockfile_file_1 = require("@pnpm/lockfile-file");
const modules_yaml_1 = require("@pnpm/modules-yaml");
const normalize_registries_1 = __importDefault(require("@pnpm/normalize-registries"));
const path = require("path");
const realpathMissing = require("realpath-missing");
async function default_1(projects, opts) {
    var _a, _b, _c, _d, _e;
    const relativeModulesDir = (_a = opts.modulesDir) !== null && _a !== void 0 ? _a : 'node_modules';
    const rootModulesDir = await realpathMissing(path.join(opts.lockfileDir, relativeModulesDir));
    const modules = await modules_yaml_1.read(rootModulesDir);
    return {
        currentHoistPattern: modules === null || modules === void 0 ? void 0 : modules.hoistPattern,
        currentPublicHoistPattern: modules === null || modules === void 0 ? void 0 : modules.publicHoistPattern,
        hoist: !modules ? undefined : Boolean(modules.hoistPattern),
        hoistedDependencies: (_b = modules === null || modules === void 0 ? void 0 : modules.hoistedDependencies) !== null && _b !== void 0 ? _b : {},
        include: (_c = modules === null || modules === void 0 ? void 0 : modules.included) !== null && _c !== void 0 ? _c : { dependencies: true, devDependencies: true, optionalDependencies: true },
        modules,
        pendingBuilds: (_d = modules === null || modules === void 0 ? void 0 : modules.pendingBuilds) !== null && _d !== void 0 ? _d : [],
        projects: await Promise.all(projects.map(async (project) => {
            var _a, _b;
            const modulesDir = await realpathMissing(path.join(project.rootDir, (_a = project.modulesDir) !== null && _a !== void 0 ? _a : relativeModulesDir));
            const importerId = lockfile_file_1.getLockfileImporterId(opts.lockfileDir, project.rootDir);
            return {
                ...project,
                binsDir: (_b = project.binsDir) !== null && _b !== void 0 ? _b : path.join(project.rootDir, relativeModulesDir, '.bin'),
                id: importerId,
                modulesDir,
            };
        })),
        registries: (modules === null || modules === void 0 ? void 0 : modules.registries) && normalize_registries_1.default(modules.registries),
        rootModulesDir,
        skipped: new Set((_e = modules === null || modules === void 0 ? void 0 : modules.skipped) !== null && _e !== void 0 ? _e : []),
    };
}
exports.default = default_1;
