// @pnpm/outdated/lib/createManifestGetter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getManifest = exports.createManifestGetter = void 0;
const client_1 = require("@pnpm/client");
const pick_registry_for_package_1 = __importDefault(require("@pnpm/pick-registry-for-package"));
function createManifestGetter(opts) {
    const resolve = client_1.createResolver({ ...opts, authConfig: opts.rawConfig });
    return getManifest.bind(null, resolve, opts);
}
exports.createManifestGetter = createManifestGetter;
async function getManifest(resolve, opts, packageName, pref) {
    var _a;
    const resolution = await resolve({ alias: packageName, pref }, {
        lockfileDir: opts.lockfileDir,
        preferredVersions: {},
        projectDir: opts.dir,
        registry: pick_registry_for_package_1.default(opts.registries, packageName),
    });
    return (_a = resolution === null || resolution === void 0 ? void 0 : resolution.manifest) !== null && _a !== void 0 ? _a : null;
}
exports.getManifest = getManifest;
