// pnpm/lib/parseCliArgs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const parse_cli_args_1 = __importDefault(require("@pnpm/parse-cli-args"));
const cmd_1 = require("./cmd");
const shorthands_1 = __importDefault(require("./shorthands"));
const RENAMED_OPTIONS = {
    'lockfile-directory': 'lockfile-dir',
    prefix: 'dir',
    'shrinkwrap-directory': 'lockfile-dir',
    store: 'store-dir',
};
function parseCliArgs(inputArgv) {
    return parse_cli_args_1.default({
        fallbackCommand: 'run',
        getCommandLongName: cmd_1.getCommandFullName,
        getTypesByCommandName: cmd_1.getCliOptionsTypes,
        renamedOptions: RENAMED_OPTIONS,
        shorthandsByCommandName: cmd_1.shorthandsByCommandName,
        universalOptionsTypes: cmd_1.GLOBAL_OPTIONS,
        universalShorthands: shorthands_1.default,
    }, inputArgv);
}
exports.default = parseCliArgs;
