// @pnpm/lockfile-utils/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPkgShortId = exports.satisfiesPackageManifest = exports.pkgSnapshotToResolution = exports.packageIsIndependent = exports.packageIdFromSnapshot = exports.nameVerFromPkgSnapshot = void 0;
__exportStar(require("@pnpm/lockfile-types"), exports);
const nameVerFromPkgSnapshot_1 = require("./nameVerFromPkgSnapshot");
exports.nameVerFromPkgSnapshot = nameVerFromPkgSnapshot_1.default;
const packageIdFromSnapshot_1 = require("./packageIdFromSnapshot");
exports.packageIdFromSnapshot = packageIdFromSnapshot_1.default;
const packageIsIndependent_1 = require("./packageIsIndependent");
exports.packageIsIndependent = packageIsIndependent_1.default;
const pkgSnapshotToResolution_1 = require("./pkgSnapshotToResolution");
exports.pkgSnapshotToResolution = pkgSnapshotToResolution_1.default;
const satisfiesPackageManifest_1 = require("./satisfiesPackageManifest");
exports.satisfiesPackageManifest = satisfiesPackageManifest_1.default;
// for backward compatibility
const dependency_path_1 = require("dependency-path");
exports.getPkgShortId = dependency_path_1.refToRelative;
