// pnpm/lib/cmd/bin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.help = exports.commandNames = exports.cliOptionsTypes = exports.rcOptionsTypes = void 0;
const config_1 = require("@pnpm/config");
const R = require("ramda");
const renderHelp = require("render-help");
exports.rcOptionsTypes = cliOptionsTypes;
function cliOptionsTypes() {
    return R.pick([
        'global',
    ], config_1.types);
}
exports.cliOptionsTypes = cliOptionsTypes;
exports.commandNames = ['bin'];
function help() {
    return renderHelp({
        description: 'Print the directory where pnpm will install executables.',
        descriptionLists: [
            {
                title: 'Options',
                list: [
                    {
                        description: 'Print the global executables directory',
                        name: '--global',
                        shortAlias: '-g',
                    },
                ],
            },
        ],
        usages: ['pnpm bin [-g]'],
    });
}
exports.help = help;
async function handler(opts) {
    return opts.bin;
}
exports.handler = handler;
